### DISCUSSION TITLE
Any ability to use web camera STILL image mode to take picture?

### DISCUSSION THREAD

**Discussion by jdredd87** [FactoryXCode/MfPack#24] [Q&A] [answered]
I want to take a still image.  Not just save a random frame from a video stream.

Using the Windows Media Foundation API ( not DirectX )

If you use Windows 10 Camera app, say on a surface pro.
The preview mode is 1920x1080 mode.
But when you take the STILL image, it is 3264x2448 ( 8 MP )

Downloaded latest MF pack and don't see this ability. So not sure if it is there and missing something.
If it does, can a demo be added?

**Comment by FactoryXCode** [comment]
Hello, 

Depending on the hardware a still image, snap-shot or freeze-frame kan be taken from it's still image pin. For this you have to use DirectShow. [Capturing an Image From a Still Image Pin](https://docs.microsoft.com/en-us/windows/win32/directshow/capturing-an-image-from-a-still-image-pin)

Media Foundation does have this ability.

There are 2 scenarios:

1# In combination with the WIC (which is also part of MfPack), it is possible to pick a frame from the videostream and convert this frame to the proper format. [as been suggested here](https://social.msdn.microsoft.com/Forums/windowsdesktop/en-US/7d2686cc-bba4-445b-a9af-2bc9ca804cb1/how-to-use-media-foundation-to-capture-a-webcam-shot-still-to-image-file?forum=mediafoundationdevelopment)

A good readable "how to" can be found [here on Code Project](https://www.codeproject.com/Articles/776058/Capturing-Live-video-from-Web-camera-on-Windows-an) 

2# By using the [IMFCapturePhotoSink interface](https://docs.microsoft.com/en-us/windows/win32/api/mfcaptureengine/nn-mfcaptureengine-imfcapturephotosink) (WinApi.MediaFoundationApi.MfCaptureEngine.pas), you can extract still images from a video stream, as explained [here](https://docs.microsoft.com/en-us/windows/win32/api/mfcaptureengine/nf-mfcaptureengine-imfcaptureengine-takephoto).

Greetings, Tony.

**Comment by Ciaran3** [comment]
I might look into doing a sample project demonstrating webcam to still image capture. It seems like one of the few remaining use cases that is not covered.

**Reply by FactoryXCode** [reply]
That would be great.
But, first update your version, because a lot of code has been updated and/or fixed.

**Reply by Ciaran3** [reply]
Thanks, I will do. Good work with all the updates. I haven't touched mine recently because of the 'if it ain't broke' rule :)

**Reply by FactoryXCode** [reply]
I'm working on a MfCaptureEngine sample using the IMFCapturePhotoSink interface.

**Reply by Ciaran3** [reply]
> I'm working on a MfCaptureEngine sample using the IMFCapturePhotoSink interface.

Nice, I would be interested how it goes.

I've spent a lot of time working on an example that uses the IMFSourceReader.
https://github.com/Ciaran3/MfPack/tree/Master/MfPack/SandBox/CameraFrameCapture

The downside is no 'live' preview is visible. Although the captured images can be displayed instead.
The priority I was working on was speed. I've managed to capture up to 21 fps from a 4K camera. If the CapturePhotoSink is able to do that same capture rate at full resolution, then its probably a better solution - or just an alternative demonstration of preview and capture. The part that is slowing down my UI is the image preview - at that resolution and frame rate, the painting of each image consumes most of the CPU time.

**Reply by FactoryXCode** [reply]
Ok, I'm progressed this far:
I get a preview and option to send the snapshot to file, hwnd destination(any control that has a THandle ( the TPaintBox doesn't have one?) or stream. 
After posting a message the iSample (like what your sample does) needs to be processed further (or not, because I think the code behind the interface will take care of that). I did not come to this point yet.

I pushed the sample to Sandbox, the code has some errors, but gives some impression.

[This is the Sample](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/SandBox/MFCaptureEngineVideoCapture)

Please feel free to join the code of this sample!

**Reply by Ciaran3** [reply]
Thanks for that Tony, it looks like a good start.
I had a few compilation errors (madExcept and in mnuTakePhotoClick), but got it compiling.
If you need code for the resolution selection feel free to re-use some of the 'PopulateStreamFormats' code from here: 
https://github.com/Ciaran3/MfPack/blob/Master/MfPack/SandBox/CameraFrameCapture/CameraCapture.pas

**Reply by FactoryXCode** [reply]
Hi,
This far I modified and added SampleConverter.pas to send the IMFSample as bitmap to TPaintbox. However, I'm questioning myself if a save to file from this control has the same quality as copying the IMFSample to for instance a IMFSample2 and holds this sample until user wants to save the photo displayed in the TPaintBox and save the snapshot by calling pPhoto.SetOutputFileName. Because calling pPhoto.SetSampleCallback will override a call to pPhoto.SetOutputFileName. :-(`

Btw: Your sample looks great!  
I have to check it for all, but why it's not possible to have a preview? It makes sense to me that a preview can't show a snapshot, but having a life preview should be possible, I think. For that you need a rendersurface that at least have a handle (the TPaintbox doesn't have one).

**Reply by Ciaran3** [reply]
I haven't looked into the IMFCaptureRecordSink, but I'm guessing saving directly from that would provide the best quality. Otherwise if you have the actual sample, you could always save that data to a memory stream and then load a BMP from that at a later stage.

With the previous method I was using, it seemed to have an exclusive lock on the camera - For example, opening an other camera app will display an 'in use error'. I've just added an option for DirectX decoding to put some of the load on the GPU instead of CPU. With that, I think I can share the render surface. Not sure if I will spend the time on that.

By the way, for some reason the DirectX rendering provides a much darker image.. You can see the difference in my project by switching on and off the DirectX option.. Maybe the ISO speed or brightness needs to be adjusted? I'll have to look into it.
Strange - The Win 10 'Camera' app shows the same thing- just increasing the brightness slider from zero renders a clearer image.

PS- Thanks! A serious number of hours has gone into it.. It always amazes me how few examples exist with the amount of functionality that is in the MfPack projects.. At least not many people sharing ;) So much time going through the Ms doco trying to find the most efficient method.

**Reply by FactoryXCode** [reply]
> With the previous method I was using, it seemed to have an exclusive lock on the camera 

Well you did not implement the OnDeviceChange callback. For cameras or other plug-in devices this is a must.

> By the way, for some reason the DirectX rendering provides a much darker image

This I noticed too. Maybe somewhere there is a default in the API. I did not discover yet.

> how few examples exist with the amount of functionality that is in the MfPack projects

Well, the API is not commonly known as DirectShow is. Like DSPack (DirectShow) had a great audience until the translators felt like Embarcadero stole their efforts..  And that was the end of this very useful site/forum. In that period we (Peter and I)  decided to continue the translations of the DSPack (DirectShow) successor MFPack (Media Foundation). 
Another fact is, it's a difficult to understand API for most programmers, because , I think , the docs from Microsoft are a little confusing about explaining the API (although I think they'r getting better). They made some examples in C++ language (The classic samples), but they are not very consistent with their cpp code on this point since the introduction of MS new compilers (and language).

**Reply by Ciaran3** [reply]
> Well you did not implement the OnDeviceChange callback. For cameras or other plug-in devices this is a must.

ah thanks, hopefully its as simple as that!

> This I noticed too. Maybe somewhere there is a default in the API. I did not discover yet.

I'll let you know if I figure it out. Its bugging me as it really makes the DirectX version look worse in most cases vs software decoding.

I agree, most of the Microsoft samples are very limited and often miss out the critical information on how to get it working (well). Hopefully the fully functional MFPack projects are saving a lot of people time, even if they aren't using Delphi. I'm not picky, any C++ Media Foundation projects I find, I look through for anything I might be missing :)

**Reply by FactoryXCode** [reply]
> I look through for anything I might be missing :)

Same here ..

**Reply by Ciaran3** [reply]
Re: Brightness - So far I've only found [DXVA2_ProcAmp_Brightness](https://docs.microsoft.com/en-us/windows/win32/medfound/procamp-settings) . It works fine when you use that to increase the brightness but it doesn't feel like the correct solution. Back to searching...

**Reply by FactoryXCode** [reply]
I'm asking myself why my camera on your example exposes 11 formats and mine about 33?

**Reply by Ciaran3** [reply]
> I'm asking myself why my camera on your example exposes 11 formats and mine about 33?

Is it the check 'IsFormatAvailable' in CameraCapture.pas? That filters out lower frame rates and formats that are not supported for conversion.

**Reply by FactoryXCode** [reply]
> Is it the check 'IsFormatAvailable' in CameraCapture.pas?

I'm still getting 6 more.

Another thing: a-synchronous operations have a thread priority of THREAD_PRIORITY_NORMAL. On high CPU usage  the thread of  an a-synchronous operation can take some while. That means, a calculation for example  memory usage or processing a sample may lead to a miscalculation.  Therefore it would be a good idea - I think - to give those another thread priority than the default. See [here](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setthreadpriority#:~:text=The%20SetThreadPriority%20function%20enables%20setting,base%20priority%20level%20to%206.)

I don't know if Delhi has a class that is able to get or set thread priorities. Otherwise I will write one.

**Reply by Ciaran3** [reply]
Thanks for that suggestion regarding thread priority - it sounds like a good idea. Although the sample reader appears to spawn multiple threads to read samples, so i'm not sure it would be efficient setting the priority on each new thread? 
Ideally, everything should be outside of the main thread - as it is possible with the current code for the main thread to become unresponsive when using 'burst mode' (continuous capture).

**Comment by FactoryXCode** [comment]
> The downside is no 'live' preview is visible. Although the captured images can be displayed instead.

This can be done by implementing the TCPreview class from sample MfCaptureD3D9.  Here the class will show the preview processed by IMFSourceReaderCallback.OnReadSample. The only thing you have to do is create a Window class like for instance TPanel to show the preview.

I'll try to implement this.

**Comment by FactoryXCode** [comment]
My suggestion is: To adjust camera properties ( like brightness, format etc.) a sink should be implemented in OnReadSample, when a sample is token.  Therefore you should use the interfaces from WinApi.MediaFoundationApi.CameraUIControl.pas.  I think that is more likely the way to implement a preview and conversion for supported formats  see [here](https://docs.microsoft.com/en-us/windows/win32/medfound/video-processor-mft)

**Comment by FactoryXCode** [comment]
OK, I'd wrote a function that doesn't need a session or d3d9 surface to get a preview. I did not test it yet, but I think in theory it should work with your sample.

```
// This method returns a IMFActivate that is able to render to a window or any other visual component that has a THandle (HWND)
function CreateMediaSinkActivateWindowRenderer(pSourceSD: IMFStreamDescriptor;
                                                                                   dwStream: DWORD; 
                                                                                   hVideoWindow: HWND;  // Handle to the video clipping window.
                                                                                   pActivate: IMFActivate): HRESULT;
var
  hr: HResult;
  pmtHandler: IMFMediaTypeHandler;
  pmfActivate: IMFActivate;
  guidMajorType: TGUID;
  pmfStreamSink: IMFStreamSink;
  pmfVideoSink: IMFMediaSink;
  dwID: DWord;

label
  Done;

begin

  // Get the media type handler for the stream.
  hr := pSourceSD.GetMediaTypeHandler(pmtHandler);
  if FAILED(hr) then
    goto done;

  // Get the major media type.
  hr := pmtHandler.GetMajorType(guidMajorType);
  if FAILED(hr) then
    goto Done;

  if (MFMediaType_Video = guidMajorType) then
    begin
      // Create the video renderer.
      hr := MFCreateVideoRendererActivate(hVideoWindow,
                                                                    pActivate);

      hr := pActivate.ActivateObject(IID_IMFMediaSink,
                                                       Pointer(pmfVideoSink));
      if SUCCEEDED(hr) then
        begin
          hr := pmfVideoSink.AddStreamSink(dwStream,
                                                                   nil,
                                                                   pmfStreamSink);
          if SUCCEEDED(hr) then
            begin
              hr := pmfStreamSink.GetIdentifier(dwID);
              if SUCCEEDED(hr) then
                SafeRelease(pmfStreamSink);
            end;
        end;
    end
  else
      begin
        // Invalid or no stream present
        hr := MF_E_CAPTURE_SOURCE_NO_VIDEO_STREAM_PRESENT;
      end;

  if FAILED(hr) then
    goto Done;

  pActivate := pmfActivate;

done:
  Result := hr;
end;

```

**Reply by FactoryXCode** [reply]
I checked your sample, but getting a MFVideoFormat_YUY2 format the frame is upside down. the MFVideoFormat_NV12 format is doing well.

**Reply by Ciaran3** [reply]
> I checked your sample, but getting a MFVideoFormat_YUY2 format the frame is upside down. the MFVideoFormat_NV12 format is doing well.

Ah yes, now I remember why I had the negative value for BITMAPINFOHEADER height. I have added that format to an array 'FTopDownFormats' and a 'IsTopDown' function. I'm not 100% sure what other formats provide top-down data - so that array might need to be added to.

I did have a go at adding the Preview with the code you provided previously, but had no luck. I think it involves a fair bit more work and I'm not sure when I'll get to that.

**Reply by FactoryXCode** [reply]
I'll give it a try.

**Comment by FactoryXCode** [comment]
I wonder why there is a function IsInputSupported and why the list of supported formats is limited to the list in function SetSupportedInputs(); 
As I read MfApi.pas (See from line 2610) the list of MF supported formats exceeds the formats defined in SetSupportedInputs().
Maybe the list is a bit behind the MF updates of the last past years?

**Reply by Ciaran3** [reply]
> MfApi

Hi Tony, thanks for your recent changes / fixes on this project.
Those formats are the only ones that are supported by the Video Processor, according to this:
https://docs.microsoft.com/en-us/windows/win32/medfound/video-processor-mft#input-formats
Which is needed to convert the sample to RGB.
I probably should make it clearer by adding a comment.

**Comment by FactoryXCode** [comment]
Thanks.
I missed that info.
A new APi update is underway: SDK 10.0.22621.0 for windows 11.

**Comment by jdredd87** [comment]
Checking to see how things on this are progressing? 

This issue haunts my dreams for years now.  Came back up yesterday.

**Reply by Ciaran3** [reply]
Hi there, have you checked out these two sample projects?
https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/Samples/MFCaptureEngineVideoCapture
https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/Samples/MfCameraFrameCapture

**Reply by FactoryXCode** [reply]
Well, the MFCaptureEngineVideo Capture sample has an issue. When changing from camera or accessing the camera again, an error is thrown (80070005 Access Denied) Maybe someone has a clue? I asked MS why this is happening with this particular interface (IMFCaptureEngine)  and not for example the IMFSourceReader (like the MfCaptureD3D9 sample).

**Comment by FactoryXCode** [comment, marked as answer]
MFCaptureEngineVideoCapture sample has been fixed.
Also the different format on a SurfacePro is caused by it's camera pin preview format.
A preview format has always  a lower quality than the actual camera format. 
So, a clip from a preview snapshot has always  a lower quality as a result.
Therefore you should not get a snapshot from a preview, but from the actual  camera presentation.
